chainstate/tx-verifier/src/transaction_verifier/reward_distribution.rs

```diff
@@ -0,0 +1,575 @@
+        .collect();
+
+    Ok(AccountingBlockRewardUndo::new(undos))
+}
+
+fn calculate_pool_owner_reward(total_reward: Amount, pool_data: &PoolData) -> Option<Amount> {
+    (total_reward - pool_data.cost_per_block())
+        .and_then(|v| v * pool_data.margin_ratio_per_thousand().value().into())
+        .and_then(|v| v / 1000)
+        .and_then(|v| v + pool_data.cost_per_block())
+}
+
+/// The reward is distributed among delegations proportionally to their balance
+fn distribute_delegations_pos_reward<P: PoSAccountingView>(
+    accounting_adapter: &mut PoSAccountingDeltaAdapter<P>,
+    block_id: Id<Block>,
+    pool_id: PoolId,
+    total_delegations_reward: Amount,
+) -> Result<Option<Vec<PoSAccountingUndo>>, ConnectTransactionError> {
+    accounting_adapter
+        .accounting_delta()
+        .get_pool_delegations_shares(pool_id)?
+        .map(|delegation_shares| {
+            let total_delegations_balance =
+                delegation_shares.values().try_fold(Amount::ZERO, |acc, v| {
+                    (acc + *v).ok_or(ConnectTransactionError::DelegationsRewardSumFailed(
+                        block_id, pool_id,
+                    ))
+                })?;
+
+            if total_delegations_balance != Amount::ZERO {
+                let rewards_per_delegation = calculate_rewards_per_delegation(
+                    &delegation_shares,
+                    block_id,
+                    total_delegations_balance,
+                    total_delegations_reward,
+                )?;
+
+                let distribute_remainder_undo = distribute_delegations_reward_remainder(
+                    accounting_adapter,
+                    block_id,
+                    pool_id,
+                    total_delegations_reward,
+                    &rewards_per_delegation,
+                )?;
+
+                let delegation_undos = rewards_per_delegation
+                    .iter()
+                    .map(|(delegation_id, reward)| {
+                        accounting_adapter
+                            .operations(TransactionSource::Chain(block_id))
+                            .delegate_staking(*delegation_id, Amount::from_atoms(*reward))
+                            .map_err(ConnectTransactionError::PoSAccountingError)
+                    })
+                    .collect::<Result<Vec<_>, _>>()?;
+
+                Ok(distribute_remainder_undo
+                    .unwrap_or_default()
+                    .into_iter()
+                    .chain(delegation_undos.into_iter())
+                    .collect())
+            } else {
+                // if total balance of all delegations is 0 then give the reward to the pool's owner
+                let increase_pool_balance_undo = accounting_adapter
+                    .operations(TransactionSource::Chain(block_id))
+                    .increase_pool_owner_balance(pool_id, total_delegations_reward)?;
+                Ok(vec![increase_pool_balance_undo])
```

Comment:
Is having this if/else an optimization? I think in general it shouldn't make a difference. You'll have zeros that will still be valid.

Reply:
These are different execution paths: I give delegations reward to the owner if there are no delegators or their total balance is 0

Reply:
Isn't this how it has to go anyway if there are no delegators?

Reply:
hmm, no. We subtract a fixed fee and a margin. But if MPT is less than 1000 the rest would be lost. 

That's how it is with the current approach. Alternatively, we could replace MPT with a reward proportional to the pledge amount, like it is with delegators.  Then it would be as you say.

Reply:
As discussed I'll refactor the flow of the distribution to avoid else clause and make it more straightforward, effectively trying to call `increase_pledge_amount` only once